src/avcodec/bitstream.rs

```diff
@@ -0,0 +1,134 @@
+    error::{Result, RsmpegError},
+    ffi,
+    shared::PointerUpgrade,
+};
+
+wrap_ref!(AVBitStreamFilter: ffi::AVBitStreamFilter);
+
+impl AVBitStreamFilter {
+    /// Find a bitstream filter instance with it's short name.
+    ///
+    /// See [`ffi::av_bsf_get_by_name`] for more info.
+    pub fn find_by_name(name: &CStr) -> Option<AVBitStreamFilterRef> {
+        unsafe { ffi::av_bsf_get_by_name(name.as_ptr()) }
+            .upgrade()
+            .map(|x| unsafe { AVBitStreamFilterRef::from_raw(x) })
+    }
+
+    // TODO: implement ffi::av_bsf_next
+    // const AVBitStreamFilter *av_bsf_next(void **opaque);
+}
+
+wrap!(AVBSFContext: ffi::AVBSFContext);
+settable!(AVBSFContext {
+    time_base_in: ffi::AVRational
+});
+
+impl AVBSFContext {
+    /// Create a new [`AVBSFContext`] instance, allocate private data and
+    /// initialize defaults for the given [`AVBitStreamFilterRef`].
+    ///
+    /// See [`ffi::av_bsf_alloc`] for more info.
+    pub fn new(filter: &ffi::AVBitStreamFilter) -> Self {
+        let mut bsfc_raw = ptr::null_mut();
+
+        unsafe {
+            ffi::av_bsf_alloc(filter, &mut bsfc_raw);
+            AVBSFContext::from_raw(bsfc_raw.upgrade().unwrap())
+        }
+    }
+    /// You need to initialize the context before you can send/receive_packets but after you set input parameters via [`AVBSFContext::set_par_in`].
+    ///
+    /// See [`ffi::av_bsf_init`] for more info.
+    pub fn init(&mut self) {
+        unsafe {
+            // TODO: Error checking
+            ffi::av_bsf_init(self.as_mut_ptr());
+        }
+    }
+    // OPTIMIZE: We should probably just be able to .upgrade() the returned c_int
+    /// Provide input data for the bitstream filter to process. To signal the end of the stream, send an NULL packet to the filter.
+    ///
+    /// See [`ffi::av_bsf_send_packet`] for more info.
+    pub fn send_packet(&mut self, packet: &mut ffi::AVPacket) -> Result<()> {
+        // TODO: Ensure init is called first
+        match unsafe { ffi::av_bsf_send_packet(self.as_mut_ptr(), packet) } {
+            0 => Ok(()),
+            e => Err(RsmpegError::AVError(e)),
+        }
+    }
+    /// Get processed data from the bitstream filter.
+    ///
+    /// See [`ffi::av_bsf_receive_packet`] for more info.
+    pub fn receive_packet(&mut self, packet: &mut ffi::AVPacket) -> Result<()> {
+        match unsafe { ffi::av_bsf_receive_packet(self.as_mut_ptr(), packet) } {
+            0 => Ok(()),
+            e => Err(RsmpegError::AVError(e)),
+        }
+    }
+    /// See [`ffi::av_bsf_flush`] for more info.
+    pub fn flush(&mut self) {
+        unsafe {
+            ffi::av_bsf_flush(self.as_mut_ptr());
+        }
+    }
+    /// Copies `source_params` into [`ffi::AVBSFContext`]'s `par_in` field. So we only need a reference to `source_params`.
+    ///
+    /// See [`ffi::avcodec_parameters_copy`] for more info.
+    pub fn set_par_in(&mut self, source_params: &ffi::AVCodecParameters) -> Result<()> {
```

Comment:
Don't use `ffi::xxx` in public interface. Please use the `avcodec::AVCodecParameters` instead.

Reply:
Do you mean in the function parameters or in the doc?

For the function parameters I used it because it allows you to seamlessly submit both `&rmpeg::avcodec::AVCodecParams` and `&ffi::AVCodecParams`. But I can see how someone might not know that `rmpeg::avcodec::AVCodecParams` automatically derefs to `ffi::AVCodecParams` and think they **must** convert the `rmpeg::avcodec::AVCodecParams` first.

I guess it depends on how smart the assumed user is. I leave that assumption up to you.

Do you want me to change it?

Reply:
Yes, the ability of submitting both is convenient. But the wrapper is meant to not letting people touch the `fii::*`, Directly hold the `ffi::*` usually leads to unsafety. And user shouldn't touch them. So I think it's not a good idea to expose them in the public interface.

Reply:
Ok I'll change this.